lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -1272,17 +1281,108 @@ void connectToMemory(handshake::FuncOp f, MemRefToMemoryAccessOp MemRefOps,
+/// wraps the partial update function.
+/// Next, the function operation is expected to go from illegal to legalized,
+/// after matchAndRewrite returned true. To work around this,
+/// LowerFuncOpTarget::loweredFuncs is used to communicate between the target
+/// and the conversion, to indicate that the partial lowering was completed.
+template <typename TFuncOp>
+struct PartialLowerFuncOp : public ConversionPattern {
+  using PartialLoweringFunc =
+      std::function<LogicalResult(TFuncOp, ConversionPatternRewriter &)>;
+
+public:
+  PartialLowerFuncOp(LowerFuncOpTarget<TFuncOp> &target, MLIRContext *context,
+                     LogicalResult &loweringResRef,
+                     const PartialLoweringFunc &fun)
+      : ConversionPattern(TFuncOp::getOperationName(), 1, context),
+        m_target(target), loweringRes(loweringResRef), m_fun(fun) {}
+  using ConversionPattern::ConversionPattern;
+  LogicalResult
+  matchAndRewrite(Operation *funcOp, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    assert(isa<TFuncOp>(funcOp));
+    rewriter.updateRootInPlace(funcOp, [&] {
+      loweringRes = m_fun(dyn_cast<TFuncOp>(funcOp), rewriter);
+    });
+    m_target.loweredFuncs[funcOp] = true;
+    return loweringRes;
+  };
+
+private:
+  LowerFuncOpTarget<TFuncOp> &m_target;
+  LogicalResult &loweringRes;
+  PartialLoweringFunc m_fun;
+};
+
+/// Calls to rewriter.replaceOpWithNewOp are invalid, given the commitment of
+/// all side-effects of of any op replacements are deferred until after the
+/// return of matchAndRewrite. This is incompatible with the design of the
+/// StandardToHandshake pass, which converts a top-level FuncOp in a sequential
+/// manner, assuming any side-effects from operand replacements (in particular
+/// rewriting of SSA use-def's) are performed instantly.
+/// The following convenience function wraps a replacement pattern @p TConv
+/// inside a distinct call to applyPartialConversion, ensuring that any call to
+/// rewrite replacement methods have been fully performed after this call. Each
+/// replacement pattern is expected to define their own conversion target, to
+/// guide pattern matching.
+///
+template <typename TConv, typename TFuncOp, typename... TArgs>
+LogicalResult lowerToHandshake(TFuncOp op, MLIRContext *context,
+                               TArgs &... args) {
+  RewritePatternSet patterns(context);
+  auto target = LowerFuncOpTarget<TFuncOp>(*context);
+  LogicalResult partialLoweringSuccessfull = success();
+  patterns.add<TConv>(target, context, partialLoweringSuccessfull, args...);
+  return success(
+      applyPartialConversion(op, target, std::move(patterns)).succeeded() &&
+      partialLoweringSuccessfull.succeeded());
+}
+
+// Convenience function for running lowerToHandshake with a partial
+// handshake::FuncOp lowering function.
+template <typename TFuncOp = handshake::FuncOp>
```

Comment:
I think the default here is unnecessary, since TFuncOp exists in the argument signature?  I think template deduction should work in this case?